core/consensus/babe/src/lib.rs

```diff
@@ -95,8 +203,845 @@ impl<Hash> CompatibleDigestItem for generic::DigestItem<Hash, Public, Secret> {
 	/// If this item is an BABE seal, return the slot number and signature.
 	fn as_babe_seal(&self) -> Option<BabeSeal> {
 		match self {
-			generic::DigestItem::Consensus(BABE_ENGINE_ID, seal) => Decode::decode(&mut &seal[..]),
-			_ => None,
+			generic::DigestItem::Consensus(BABE_ENGINE_ID, seal) => {
+				match Decode::decode(&mut &seal[..]) {
+					s@Some(_) => s,
+					s@None => {
+						warn!(target: "babe", "Failed to decode {:?}", seal);
```

Comment:
I'm not sure this or the other warn in this function should be `warn`. It's expected that we'll have bad peers who give us bad blocks, and this code is the first line of defence against them.